tla-bmcmt/src/main/scala/at/forsyte/apalache/tla/bmcmt/rules/SetInRuleWithArrays.scala

```diff
@@ -90,8 +91,9 @@ class SetInRuleWithArrays(rewriter: SymbStateRewriter) extends SetInRule(rewrite
           nextState = rewriter.rewriteUntilDone(nextState.setRex(tla.appFun(funCell.toNameEx, funsetDomElem.toNameEx)))
           val funAppRes = nextState.asCell
           val powSetDom = nextState.arena.getDom(funsetCdm)
-          val subsetEq = tla.subseteq(funAppRes.toNameEx, powSetDom.toNameEx)
-          nextState = rewriter.rewriteUntilDone(nextState.setRex(subsetEq))
+          // rename to obtain a unique name for the bound variable
+          val quant = renaming(tla.forall(tla.name("x"), funAppRes.toNameEx, tla.in(tla.name("x"), powSetDom.toNameEx)))
+          nextState = rewriter.rewriteUntilDone(nextState.setRex(quant))
```

Comment:
I think my previous comment got deleted. I think the proper way of doing that is by rewriting `f \in [S -> SUBSET T]` into `DOMAIN f = S /\ \A x \in S: \A y \in f[x]: y \in T` in `Keramelizer`. Since it is about rewriting expressions, we should move it in preprocessing.

Reply:
So is `f \in [S -> SUBSET T]` not in KerA+?

Reply:
we have it in KerA+. But I had this idea for some time that we should translate expressions over functions sets and powersets in the preprocessor, so we do not have to deal with them in the rewriting rules: https://github.com/informalsystems/apalache/milestone/56

If you think that it would be easier to handle this case in `Keramelizer`, go ahead.

Reply:
I asked because you said above
> I think the proper way of doing that is by rewriting ... into ... in Keramelizer

So is the sentiment that we should do it in preprocessing, but not in Kereamelizer?

Reply:
We have `Keramelizer` and `ExprOptimizer`. `Keramelizer` was designed for general reductions, whereas `ExprOptimizer` was designed for sound but ad hoc simplifications. Your case looks like a fit for `Keramelizer`.

Reply:
Okay. Still, shouldn't that be an additional preprocessor then? It seems confusing to have a class `Keramelizer` with its Scaladoc saying `a simplifier from TLA+ to KerA+` rewrite an expression that is already KerA+.

Reply:
Also, for my own understanding: why is it better to have this in Keramlizer than the rewrite rule?

Reply:
I mean, KerA+ is not set in stone. We have removed some expressions from it in the past. It would be better to handle it in the preprocessor than in a rewrite rule, since preprocessing is purely about TLA-to-TLA transformations, whereas rewriting rules are producing SMT constraints and cells. So having a purely TLA-to-TLA transformation in a rewriting rule usually means that this transformation should be moved up in the pipeline. Normally, TLA-to-TLA transformations do not depend on the SMT, whereas rewriting rules heavily depend on the encoding.

Reply:
> Okay. Still, shouldn't that be an additional preprocessor then? It seems confusing to have a class `Keramelizer` with its Scaladoc saying `a simplifier from TLA+ to KerA+` rewrite an expression that is already KerA+.

In principle, you are right. I wanted to do some systematic refactoring of the preprocessing passes, e.g, see #480. I would do that in Keramelizer at the moment and come back to a more principled solution later, when we understand what we want from the TLA-to-TLA framework.

Reply:
Everything about moving this into Keramelizer seems to be orthogonal to this PR. I'd like to understand what's wrong with the proposed solution. Are you concerned about the name generation or the quantifier? The OOPSLA rule also expands this expression into a conjunction over the function relation.

CC @rodrigo7491

Reply:
Yes, I am concerned about name generation in the rewriting rule.

Reply:
Good point! @rodrigo7491, do you see any (performance) problem with rewriting this in Keramelizer as suggested by Igor?
One thing to keep in mind is that also `Apalache!SelectInSet` goes through [SetInRuleWithArrays.scala](https://github.com/informalsystems/apalache/pull/1716/files/83bbac46be1236ee55ee957c89c8269465351204#diff-5624951d59d02e27a1efb359513855166e609fcfd97e454c67a72d539fa246f0)

Reply:
From what I understand we will get a quantified expression in the end, regardless of where the rewriting happens, so I don't think it should be a problem. I don't get your point about the `SelectInSet` IR op though @thpani.

Reply:
> I don't get your point about the `SelectInSet` IR op though @thpani.

Ah, I just meant that `SelectInSet` won't be rewritten by `Keramelizer`.